python-rx fails to build with Python 3.11: AttributeError: module 'asyncio' has no attribute 'coroutine

https://bugzilla.redhat.com/show_bug.cgi?id=2019856

python-rx fails to build with Python 3.11.0a1.

  Traceback (most recent call last):
    File "/usr/lib64/python3.11/threading.py", line 1031, in _bootstrap_inner
      self.run()
      ^^^^^^^^^^
    File "/usr/lib64/python3.11/threading.py", line 968, in run
      self._target(*self._args, **self._kwargs)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py", line 113, in thread_target
      @asyncio.coroutine
       ^^^^^^^^^^^^^^^^^
  AttributeError: module 'asyncio' has no attribute 'coroutine'
  
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action
  /builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py:27: DeprecationWarning: There is no current event loop
    loop = asyncio.get_event_loop()

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_cancel
  /builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py:74: DeprecationWarning: There is no current event loop
    loop = asyncio.get_event_loop()

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_cancel_same_loop
  /usr/lib/python3.11/site-packages/_pytest/threadexception.py:75: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-116 (thread_target)
  
  Traceback (most recent call last):
    File "/usr/lib64/python3.11/threading.py", line 1031, in _bootstrap_inner
      self.run()
      ^^^^^^^^^^
    File "/usr/lib64/python3.11/threading.py", line 968, in run
      self._target(*self._args, **self._kwargs)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py", line 113, in thread_target
      @asyncio.coroutine
       ^^^^^^^^^^^^^^^^^
  AttributeError: module 'asyncio' has no attribute 'coroutine'
  
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_cancel_same_thread
  /usr/lib/python3.11/site-packages/_pytest/threadexception.py:75: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-117 (thread_target)
  
  Traceback (most recent call last):
    File "/usr/lib64/python3.11/threading.py", line 1031, in _bootstrap_inner
      self.run()
      ^^^^^^^^^^
    File "/usr/lib64/python3.11/threading.py", line 968, in run
      self._target(*self._args, **self._kwargs)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py", line 113, in thread_target
      @asyncio.coroutine
       ^^^^^^^^^^^^^^^^^
  AttributeError: module 'asyncio' has no attribute 'coroutine'
  
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_due
  /builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py:49: DeprecationWarning: There is no current event loop
    loop = asyncio.get_event_loop()

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_now
  /builddir/build/BUILD/RxPY-3.2.0/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py:13: DeprecationWarning: There is no current event loop
    loop = asyncio.get_event_loop()

tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_now_units
  /usr/lib64/python3.11/unittest/case.py:646: DeprecationWarning: It is deprecated to return a value!=None from a test case (<bound method TestAsyncIOThreadSafeScheduler.test_asyncio_threadsafe_schedule_now_units of <tests.test_scheduler.test_eventloop.test_asynciothreadsafescheduler.TestAsyncIOThreadSafeScheduler testMethod=test_asyncio_threadsafe_schedule_now_units>>)
    return self.run(*args, **kwds)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_observable/test_fromfuture.py::TestFromFuture::test_future_cancel
FAILED tests/test_observable/test_fromfuture.py::TestFromFuture::test_future_dispose
FAILED tests/test_observable/test_fromfuture.py::TestFromFuture::test_future_failure
FAILED tests/test_observable/test_fromfuture.py::TestFromFuture::test_future_success
FAILED tests/test_observable/test_start.py::TestStart::test_start_async - Att...
FAILED tests/test_observable/test_start.py::TestStart::test_start_async_error
FAILED tests/test_scheduler/test_eventloop/test_asyncioscheduler.py::TestAsyncIOScheduler::test_asyncio_schedule_action
FAILED tests/test_scheduler/test_eventloop/test_asyncioscheduler.py::TestAsyncIOScheduler::test_asyncio_schedule_action_cancel
FAILED tests/test_scheduler/test_eventloop/test_asyncioscheduler.py::TestAsyncIOScheduler::test_asyncio_schedule_action_due
FAILED tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action
FAILED tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_cancel
FAILED tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_cancel_same_loop
FAILED tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py::TestAsyncIOThreadSafeScheduler::test_asyncio_threadsafe_schedule_action_due
========== 13 failed, 1309 passed, 10 skipped, 27 warnings in 18.75s ===========

The @asyncio.coroutine decorator enabling legacy generator-based coroutines to be compatible with async/await code. The function has been deprecated since Python 3.8 and the removal was initially scheduled for Python 3.10. Use async def instead. (Contributed by Illia Volochii in bpo-43216.)

https://bugs.python.org/issue43216
https://docs.python.org/3.11/whatsnew/3.11.html

For the build logs, see:
https://copr-be.cloud.fedoraproject.org/results/@python/python3.11/fedora-rawhide-x86_64/02910611-python-rx/

For all our attempts to build python-rx with Python 3.11, see:
https://copr.fedorainfracloud.org/coprs/g/python/python3.11/package/python-rx/

Testing and mass rebuild of packages is happening in copr. You can follow these instructions to test locally in mock if your package builds with Python 3.11:
https://copr.fedorainfracloud.org/coprs/g/python/python3.11/

Let us know here if you have any questions.

Python 3.11 is planned to be included in Fedora 37. To make that update smoother, we're building Fedora packages with all pre-releases of Python 3.11.
A build failure prevents us from testing all dependent packages (transitive [Build]Requires), so if this package is required a lot, it's important for us to get it fixed soon.
We'd appreciate help from the people who know this package best, but if you don't want to work on this now, let us know so we can try to work around it on our side.

--- a/examples/asyncio/toasyncgenerator.py
+++ b/examples/asyncio/toasyncgenerator.py
@@ -37,8 +37,7 @@ def to_async_generator(sentinel=None):
 
         source.pipe(ops.materialize()).subscribe(on_next)
 
-        @asyncio.coroutine
-        def gen():
+        async def gen():
             """Generator producing futures"""
             nonlocal future
 
@@ -50,8 +49,7 @@ def to_async_generator(sentinel=None):
     return _to_async_generator
 
 
-@asyncio.coroutine
-def go(loop):
+async def go(loop):
     scheduler = AsyncIOScheduler(loop)
 
     xs = rx.from_([x for x in range(10)], scheduler=scheduler)
@@ -60,7 +58,7 @@ def go(loop):
     # Wish we could write something like:
     # ys = (x for x in yield from gen())
     while True:
-        x = yield from gen()
+        x = await gen()
         if x is None:
             break
         print(x)

--- a/tests/test_observable/test_fromfuture.py
+++ b/tests/test_observable/test_fromfuture.py
@@ -11,8 +11,7 @@ class TestFromFuture(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [False, True, False]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             future = Future()
             future.set_result(42)
 
@@ -36,8 +35,7 @@ class TestFromFuture(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [True, False, True]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             error = Exception('woops')
 
             future = Future()
@@ -63,8 +61,7 @@ class TestFromFuture(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [True, False, True]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             future = Future()
             source = rx.from_future(future)
 
@@ -87,8 +84,7 @@ class TestFromFuture(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [True, True, True]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             future = Future()
             future.set_result(42)
 

--- a/tests/test_observable/test_start.py
+++ b/tests/test_observable/test_start.py
@@ -20,8 +20,7 @@ class TestStart(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [False]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             def func():
                 future = Future()
                 future.set_result(42)
@@ -40,8 +39,7 @@ class TestStart(unittest.TestCase):
         loop = asyncio.get_event_loop()
         success = [False]
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             def func():
                 future = Future()
                 future.set_exception(Exception(str(42)))
@@ -73,7 +71,6 @@ class TestStart(unittest.TestCase):
             on_completed(200)]
 
         assert(done)
-
 
     def test_start_func2(self):
         scheduler = TestScheduler()

--- a/tests/test_scheduler/test_eventloop/test_asyncioscheduler.py
+++ b/tests/test_scheduler/test_eventloop/test_asyncioscheduler.py
@@ -25,8 +25,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
     def test_asyncio_schedule_action(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             scheduler = AsyncIOScheduler(loop)
             ran = False
 
@@ -36,7 +35,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
 
             scheduler.schedule(action)
 
-            yield from asyncio.sleep(0.1)
+            await asyncio.sleep(0.1)
             assert ran is True
 
         loop.run_until_complete(go())
@@ -44,8 +43,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
     def test_asyncio_schedule_action_due(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             scheduler = AsyncIOScheduler(loop)
             starttime = loop.time()
             endtime = None
@@ -56,7 +54,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
 
             scheduler.schedule_relative(0.2, action)
 
-            yield from asyncio.sleep(0.3)
+            await asyncio.sleep(0.3)
             assert endtime is not None
             diff = endtime - starttime
             assert diff > 0.18
@@ -66,8 +64,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
     def test_asyncio_schedule_action_cancel(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             ran = False
             scheduler = AsyncIOScheduler(loop)
 
@@ -78,7 +75,7 @@ class TestAsyncIOScheduler(unittest.TestCase):
             d = scheduler.schedule_relative(0.05, action)
             d.dispose()
 
-            yield from asyncio.sleep(0.3)
+            await asyncio.sleep(0.3)
             assert ran is False
 
         loop.run_until_complete(go())

--- a/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py
+++ b/tests/test_scheduler/test_eventloop/test_asynciothreadsafescheduler.py
@@ -26,8 +26,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
     def test_asyncio_threadsafe_schedule_action(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             scheduler = AsyncIOThreadSafeScheduler(loop)
             ran = False
 
@@ -40,7 +39,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
 
             threading.Thread(target=schedule).start()
 
-            yield from asyncio.sleep(0.1)
+            await asyncio.sleep(0.1)
             assert ran is True
 
         loop.run_until_complete(go())
@@ -48,8 +47,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
     def test_asyncio_threadsafe_schedule_action_due(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             scheduler = AsyncIOThreadSafeScheduler(loop)
             starttime = loop.time()
             endtime = None
@@ -63,7 +61,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
 
             threading.Thread(target=schedule).start()
 
-            yield from asyncio.sleep(0.3)
+            await asyncio.sleep(0.3)
             assert endtime is not None
             diff = endtime - starttime
             assert diff > 0.18
@@ -73,8 +71,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
     def test_asyncio_threadsafe_schedule_action_cancel(self):
         loop = asyncio.get_event_loop()
 
-        @asyncio.coroutine
-        def go():
+        async def go():
             ran = False
             scheduler = AsyncIOThreadSafeScheduler(loop)
 
@@ -88,7 +85,7 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
 
             threading.Thread(target=schedule).start()
 
-            yield from asyncio.sleep(0.3)
+            await asyncio.sleep(0.3)
             assert ran is False
 
         loop.run_until_complete(go())
@@ -110,9 +107,8 @@ class TestAsyncIOThreadSafeScheduler(unittest.TestCase):
 
             test_body(scheduler, action, update_state)
 
-            @asyncio.coroutine
-            def go():
-                yield from asyncio.sleep(0.2)
+            async def go():
+                await asyncio.sleep(0.2)
 
             loop.run_until_complete(go())
 

